internal/csi/iam.go

```diff
@@ -0,0 +1,310 @@
+           "Effect": "Allow",
+           "Resource": "*"
+       },
+       {
+           "Effect": "Allow",
+           "Action": "elasticloadbalancing:*",
+           "Resource": "*"
+       },
+       {
+           "Effect": "Allow",
+           "Action": "cloudwatch:*",
+           "Resource": "*"
+       },
+       {
+           "Effect": "Allow",
+           "Action": "autoscaling:*",
+           "Resource": "*"
+       },
+       {
+           "Effect": "Allow",
+           "Action": "iam:CreateServiceLinkedRole",
+           "Resource": "*",
+           "Condition": {
+               "StringEquals": {
+                   "iam:AWSServiceName": [
+                       "autoscaling.amazonaws.com",
+                       "ec2scheduled.amazonaws.com",
+                       "elasticloadbalancing.amazonaws.com",
+                       "spot.amazonaws.com",
+                       "spotfleet.amazonaws.com"
+                   ]
+               }
+           }
+       }
+   ]
+}`
+)
+
+type iamResources struct {
+	svc             *iam.IAM
+	instanceProfile *iamResource
+	policy          *iamResource
+	role            *iamResource
+
+	lg *zap.Logger
+}
+
+type iamResource struct {
+	name string
+	arn  string
+}
+
+func createIAM(awsRegion string) (*iam.IAM, error) {
+	sess, err := session.NewSession(&aws.Config{Region: aws.String(awsRegion)})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a new session (%v)", err)
+	}
+	return iam.New(sess), nil
+}
+
+// awsRegion must be a valid AWS region for ec2 instances.
+// For a complete list, see entries under "Region" on the table "Amazon Elastic Compute Cloud (Amazon EC2)":
+// https://docs.aws.amazon.com/general/latest/gr/rande.html#ec2_region
+func createIAMResources(awsRegion string) (resources *iamResources, err error) {
+	resources = &iamResources{}
+
+	defer func() {
+		// Delay is needed to ensure that permissions have been propagated.
+		// See the section "Launching an Instance with an IAM Role" at
+		// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html
+		time.Sleep(10 * time.Second)
+
+		if err != nil {
+			resources.deleteIAMResources()
+		}
+	}()
+
+	// Creates new logger
+	lg, err := zap.NewProduction()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create logger (%v)", err)
+	}
+	resources.lg = lg
+
+	// Creates new session and IAM client
+	svc, err := createIAM(awsRegion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create session and IAM client (%v)", err)
+	}
+	resources.svc = svc
+	resources.lg.Info("created session and IAM client")
+
+	// AWS does not allow ':' in the names of instances for its services
+	uniqueSuffix := strings.Replace(time.Now().Format(time.RFC3339), ":", "-", -1)
+
+	// Creates instance profile
+	instanceOutput, err := resources.svc.CreateInstanceProfile(&iam.CreateInstanceProfileInput{
+		InstanceProfileName: aws.String(fmt.Sprintf("aws-k8s-tester-instance-profile-%s", uniqueSuffix)),
```

Comment:
Let's prefix without hyphens and add `"csi"` to the name

```go
InstanceProfileName: aws.String(fmt.Sprintf("awsk8stester-csi-instance-profile-%s", uniqueSuffix)),`
```

as we do:

https://github.com/aws/aws-k8s-tester/blob/e790272b0295ab27225abaf42d1b9d8381eb451e/eksconfig/config.go#L386-L392

Trying to save some characters :)

Reply:
Ok! I also changed the ending from:

```go
uniqueSuffix := strings.Replace(time.Now().Format(time.RFC3339), ":", "-", -1)
```

to:
```go
now := time.Now().UTC()
uniqueSuffix := fmt.Sprintf("%d%02d%02d%02d", now.Year(), now.Month(), now.Day(), now.Minute())
```

because I like how you did it there better

Reply:
done!